Router class implementation

ismailAde opened the pull request:
The router class gets user request through the url and breaks down the
url to determine which controller and what action is to be called as
well as parameters to pass along.
#Changes were made on a windows server

ismailAde committed:
Router class implementation

The router class gets user request through the url and breaks down the
url to determine which controller and what action is to be called as
well as parameters to pass along.
#Changes were made on a windows server

perfectmak commented on app/init.php:
```diff
@@ -0,0 +1,9 @@
+
+require_once(APP_PATH.DS.'core/Router.php');
+require_once(APP_PATH.DS.'core/Bootstrap.php');
+require_once(ROOT_PATH.DS.'vendor/autoload.php');
```
Please remove this file and use composer class loader. Look into the composer.json file for details of the class map. We are using psr-4

perfectmak commented on index.php:
```diff
@@ -1,7 +1,9 @@
 <?php
+require_once('app/init.php');
+//require(__DIR__ . '/vendor/autoload.php');
```
uncomment this and remove the app/init.php require.

perfectmak commented on index.php:
```diff
@@ -1,7 +1,9 @@
 
-use App\Core\Dispatcher;
+use App\Core\Bootstrap as Bootstrap;
+use App\Core\Router as Router;
```
You do not need the `as` statements as it is redundant here, so please remove them too

perfectmak commented on .htaccess:
```diff
@@ -0,0 +1,9 @@
+options -MultiViews
+RewriteEngine on
+
+RewriteBase /voucher-service-develop
```
RewriteBase here would prevent this codebase from being portable so remove it

perfectmak commented on app/Core/Router.php:
```diff
@@ -27,6 +33,54 @@ private function __construct() {
+        return $this->$routerInstance ?? new Router();
+    }
+
+    public function parseUrl($url = ''){
```
This function has no business here. Remove it and check out my last comment for better explanation of why.
And the same goes for the `getController()`, `getAction()` and `getParams()`

perfectmak commented:
@ismailAde This is good 👍, but the job of the router is to register routes and the callbacks that would be invoked when such routes in invoked.

So it has no business determining which controller should be used based on the url or even breaking down the url. Another class would be made to do that if required.

So it shouldn't have controller or actions information.
It should basically expose an interface that allows registering of routes and the callbacks. And also allows retrieving the callback based on the routes. So the Router should be like a Map class.

perfectmak commented:
Also @ismailAde rename the controllers directory to conform with ps-4 standards.

ismailAde committed:
Router-class

The Router class provides an interface for registering routes and their
callbacks while it also allows you to retrieve callbacks based on the
route

ismailAde commented:
Kindly go through the project now @perfectmak .

perfectmak commented on app/Core/Router.php:
```diff
@@ -29,4 +42,74 @@ private function __construct() {
+     * @param string action
+     */
+    public function get($url, $action){
+        $this->addRoute('Get', $url, $action);
```
Make the 'Get' string a class constant and reference it here, also do the same for the other types.

perfectmak commented on app/routes.php:
```diff
@@ -6,3 +6,5 @@
  * Date: 12/21/16
```
Please delete this file and this folder from this PR

ismailAde replied:
The routes.php file will be deleted now. But its in the App directory, Should the folder be deleted as well?

perfectmak commented on index.php:
```diff
@@ -1,7 +1,5 @@
 <?php
-
 use App\Core\Dispatcher;
```
Remove this use statement, it is redundant

perfectmak commented:
@ismailAde Great job. So kindly go through the comments and revert as soon as possible.

perfectmak commented:
@ismailAde Great job. So kindly go through the comments and revert as soon as possible.

ismailAde commented:
Noted, Will work on that now

ismailAde committed:
Router-Class

The Router class provides an interface for registering routes and their
callbacks while it also allows you to retrieve callbacks based on the
route.

ismailAde commented:
Done @perfectmak

perfectmak commented on app/Core/Router.php:
```diff
@@ -29,4 +51,74 @@ private function __construct() {
+     * @param string url
+     * @param string action
+     */
+    protected function addRoute($request_type='Get', $url, $action){
```
Does this line work? I mean PHP doesn't allow default parameters as the first parameter?
Which brings me to the question, have you tested this class to work?

ismailAde replied:
It skipped my mind the initial commit was not tested. However the line works. I tested and made another commit.

ismailAde committed:
Router-Class

The Router class provides an interface for registering routes and their
callbacks while it also allows you to retrieve callbacks based on the
route.

perfectmak commented on app/Core/Router.php:
```diff
@@ -29,4 +51,83 @@ private function __construct() {
+     * @param string url
+     * @param string action
+     */
+    protected function addRoute($request_type=self::GET, $url, $action){
```
This line would not work, the default argument would throw an error.
Check here: http://php.net/manual/en/functions.arguments.php for more infor. Look under incorrect usage of default function argument.

ismailAde committed:
Router-Class

Router Class implementation with PHPUnit Test
